From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: RLtools <rl-tools@localhost>
Date: Sun, 6 Sep 2026 00:00:00 +0000
Subject: [PATCH 3/8] [Support] Test for sys/resource.h before using resource
 limits

ProgramStack, Process and Program used getrlimit and setrlimit
unconditionally on Unix. Guard them with HAVE_SYS_RESOURCE_H so that a
Unix-like host without resource limits, such as WASI, takes the same
paths the non-Unix builds take: the 8 MiB default stack size, no core
file suppression, and no memory limit for child processes. WASI joins
z/OS in not warning about the missing usage times.
---
 llvm/lib/Support/ProgramStack.cpp | 4 ++--
 llvm/lib/Support/Unix/Process.inc | 6 +++++-
 llvm/lib/Support/Unix/Program.inc | 6 ++++++
 3 files changed, 13 insertions(+), 3 deletions(-)

diff --git a/llvm/lib/Support/ProgramStack.cpp b/llvm/lib/Support/ProgramStack.cpp
index 66f74ff66..cae4dcbd9 100644
--- a/llvm/lib/Support/ProgramStack.cpp
+++ b/llvm/lib/Support/ProgramStack.cpp
@@ -10,7 +10,7 @@
 #include "llvm/Config/config.h"
 #include "llvm/Support/Compiler.h"
 
-#ifdef LLVM_ON_UNIX
+#if defined(LLVM_ON_UNIX) && defined(HAVE_SYS_RESOURCE_H)
 # include <sys/resource.h> // for getrlimit
 #endif
 
@@ -42,7 +42,7 @@ uintptr_t llvm::getStackPointer() {
 }
 
 unsigned llvm::getDefaultStackSize() {
-#ifdef LLVM_ON_UNIX
+#if defined(LLVM_ON_UNIX) && defined(HAVE_SYS_RESOURCE_H)
   rlimit RL;
   getrlimit(RLIMIT_STACK, &RL);
   return RL.rlim_cur;
diff --git a/llvm/lib/Support/Unix/Process.inc b/llvm/lib/Support/Unix/Process.inc
index c6e79af44..2980efadc 100644
--- a/llvm/lib/Support/Unix/Process.inc
+++ b/llvm/lib/Support/Unix/Process.inc
@@ -19,7 +19,9 @@
 #include <optional>
 #include <fcntl.h>
 #include <sys/time.h>
+#ifdef HAVE_SYS_RESOURCE_H
 #include <sys/resource.h>
+#endif
 #include <sys/stat.h>
 #include <signal.h>
 #if defined(HAVE_MALLINFO) || defined(HAVE_MALLINFO2)
@@ -53,7 +55,7 @@ getRUsageTimes() {
   ::getrusage(RUSAGE_SELF, &RU);
   return {toDuration(RU.ru_utime), toDuration(RU.ru_stime)};
 #else
-#ifndef __MVS__ // Exclude for MVS in case -pedantic is used
+#if !defined(__MVS__) && !defined(__wasi__) // Exclude in case -pedantic is used
 #warning Cannot get usage times on this platform
 #endif
   return {std::chrono::microseconds::zero(), std::chrono::microseconds::zero()};
@@ -136,6 +138,7 @@ void Process::GetTimeUsage(TimePoint<> &elapsed,
 // their operation. To prevent the disk from filling up, this function
 // does what's necessary to prevent their generation.
 void Process::PreventCoreFiles() {
+#ifdef HAVE_SYS_RESOURCE_H
   struct rlimit rlim;
   getrlimit(RLIMIT_CORE, &rlim);
 #ifdef __linux__
@@ -158,6 +161,7 @@ void Process::PreventCoreFiles() {
   rlim.rlim_cur = 0;
 #endif
   setrlimit(RLIMIT_CORE, &rlim);
+#endif
 
 #if defined(HAVE_MACH_MACH_H) && !defined(__GNU__)
   // Disable crash reporting on Mac OS X 10.0-10.4
diff --git a/llvm/lib/Support/Unix/Program.inc b/llvm/lib/Support/Unix/Program.inc
index 489ca03b3..2aaf64ae1 100644
--- a/llvm/lib/Support/Unix/Program.inc
+++ b/llvm/lib/Support/Unix/Program.inc
@@ -29,7 +29,9 @@
 #include "llvm/Support/raw_ostream.h"
 #include <fcntl.h>
 #include <signal.h>
+#ifdef HAVE_SYS_RESOURCE_H
 #include <sys/resource.h>
+#endif
 #include <sys/stat.h>
 #if HAVE_UNISTD_H
 #include <unistd.h>
@@ -136,6 +138,7 @@ static bool RedirectIO_PS(const std::string *Path, int FD, std::string *ErrMsg,
 
 static void TimeOutHandler(int Sig) {}
 
+#ifdef HAVE_SYS_RESOURCE_H
 static void SetMemoryLimits(unsigned size) {
   struct rlimit r;
   __typeof__(r.rlim_cur) limit = (__typeof__(r.rlim_cur))(size)*1048576;
@@ -151,6 +154,7 @@ static void SetMemoryLimits(unsigned size) {
   setrlimit(RLIMIT_RSS, &r);
 #endif
 }
+#endif
 
 static std::vector<const char *>
 toNullTerminatedCStringArray(ArrayRef<StringRef> Strings, StringSaver &Saver) {
@@ -299,10 +303,12 @@ static bool Execute(ProcessInfo &PI, StringRef Program,
       }
     }
 
+#ifdef HAVE_SYS_RESOURCE_H
     // Set memory limits
     if (MemoryLimit != 0) {
       SetMemoryLimits(MemoryLimit);
     }
+#endif
 
     // Execute!
     std::string PathStr = std::string(Program);
